### ISSUE TITLE
Cannot build when using Sharp 0.33.0

### ISSUE THREAD

**Issue by Boston343** [withastro/astro#9345]
### Astro Info

```block
Astro                    v4.0.2
Node                     v18.18.2
System                   Windows (x64)
Package Manager          npm
Output                   static
Adapter                  none
Integrations             auto-import
                         @astrojs/mdx
                         @astrojs/tailwind
                         @astrojs/sitemap
                         astro-compress
```

### Describe the Bug

If I upgrade the Sharp dependency to 0.33.0 for image optimization with astro:assets, I am unable to build locally (`npm run build`). Development mode still works fine.

This happens with both Astro v3.6.4 and v4.0.2.

The error logged to the console looks like:

```bash
MissingSharp: Could not find Sharp. Please install Sharp (`sharp`) manually into your project or migrate to another image service.
    at loadSharp (file:///.../dist/chunks/astro/assets-service_x_nToqfa.mjs:352:11)
    at async Object.transform (file:///.../dist/chunks/astro/assets-service_x_nToqfa.mjs:364:15)
    at async generateImageInternal (file:///.../node_modules/astro/dist/assets/build/generate.js:119:24)
    at async generateImage (file:///.../node_modules/astro/dist/assets/build/generate.js:66:28)
    at async file:///.../node_modules/p-queue/dist/index.js:118:36 {
  loc: undefined,
  title: 'Could not find Sharp.',
  hint: "See Sharp's installation instructions for more information: https://sharp.pixelplumbing.com/install. If you are not relying on `astro:assets` to optimize, transform, or process any images, you can configure a passthrough image service instead of installing Sharp. See https://docs.astro.build/en/reference/errors/missing-sharp for more information.\n" +
    '\n' +
    'See https://docs.astro.build/en/guides/images/#default-image-service for more information on how to migrate to another image service.',
  frame: undefined,
  type: 'AstroError'
}
```

**Additional notes:** It seems to be an interaction between astro:assets and astro-compress. If I even have the line `import compress from "astro-compress"` inside `astro.config.mjs` the build will fail (I don't even have to add it to the integrations arrray). However, if I remove the compress() integration, and comment out the import statement, it will build fine. THEN I can add the astro-compress import and integration back into the integrations array and it will build and compress fine. But if I then go and add an image or otherwise cause the images to need re-generated (since they aren't cached), the issue will occur again. And then you'll have to do the comment out compress and what-not all over again.

If I downgrade sharp back to 0.32.6 it builds just fine.

### What's the expected result?

The website should build and this error should not be thrown.

### Link to Minimal Reproducible Example

[Stackblitz link](https://stackblitz.com/edit/github-ehydmu-nwlevv?file=astro.config.mjs). You need to download it, run `npm install`, then `npm run build` to see the error. See the "Additional notes" section above for more details.

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by Boston343** [comment]
> Could you provide a reproduction? I'm building many websites using Sharp 0.33.0 and Astro 4 successfully both locally and in CI with no issues.

Working on trying to replicate on Stackblitz and haven't been able to yet. Might try a local copy of the same code to see if anything changes.

**Comment by Boston343** [comment]
@Princesseuh I added a stackblitz link, although you'll need to download it to your machine and install npm packages, then attempt to build. I added an additional comment which I've copied here. I'm not sure if this is an astro problem or an astro-compress problem but it seems to be an interaction between the two.

**Additional notes:** It seems to be an interaction between astro:assets and astro-compress. If I even have the line `import compress from "astro-compress"` inside `astro.config.mjs` the build will fail (I don't even have to add it to the integrations arrray). However, if I remove the compress() integration, and comment out the import statement, it will build fine. THEN I can add the astro-compress import and integration back into the integrations array and it will build and compress fine. But if I then go and add an image or otherwise cause the images to need re-generated (since they aren't cached), the issue will occur again. And then you'll have to do the comment out compress and what-not all over again.

**Comment by josuazurbruegg** [comment]
I'm having trouble building using the most recent sharp version too. Downgrading sharp to 0.32.6 helped me as well.
I'm using Astro 4.0.3. I'm not using the astro-compress library. My project is pretty basic but I'm generating various dynamic images using the getImage server-side helper, which is the only thing slightly out of the ordinary that I'm doing.

Happy to provide a reproducible example if needed still!

P.S. this is my first project working with Astro and it has been an absolute treat! A big thank you to everyone involved in the work on the project!

**Comment by Princesseuh** [comment]
Hmm, following the reproduction steps, everything seems to work just fine for me.

One thing I do notice with the `astro-compress` stuff, is that it pulls an old version of Sharp in addition to the installed one, I wonder if there could be an incompatibility due to having multiple versions installed, perhaps. Nonetheless, I'm not able to reproduce the error in the OP 😦

**Comment by notjosh** [comment]
When I dump the actual error, I get the following exception:

```
Error: Could not load the "sharp" module using the darwin-x64 runtime
Possible solutions:
- Add platform-specific dependencies:
    npm install --os=darwin --cpu=x64 sharp
  or
    npm install --force @img/sharp-darwin-x64
```

Turns out, Sharp 0.33.0 changed how binaries are distributed (tldr: https://github.com/lovell/sharp/issues/3750#issuecomment-1751172728).

The fix for me was to run `yarn add sharp --ignore-engines`, which seems to persist even when `node_modules` is removed. It's weird though, because I already had `sharp` in `package.json`, so I'd expect it to have worked. Computers 🤷

**Comment by fflaten** [comment]
Had the same error in OP since 0.33.0 was released. Remember to clear image cache after testing older version to force image render using current version.

Using pnpm. Assumed issue was because Astro has peer dep on 0.32.*

**Comment by ppo** [comment]
I had the same problem installing my site on my server using `@astrojs/node`, using Astro 4.0.3.

Installing `"sharp": "^0.32"` on the server (with node v20.10.0) solved the problem.
On my computer (with node v21.2.0), `"sharp": "^0.33.1"` works fine.

**Comment by fflaten** [comment]
@Princesseuh 4.0.8 didn't resolve it unfortunately.

**Comment by Princesseuh** [comment]
Hmm, that's annoying. I unfortunately do not know, and I still cannot reproduce any issues myself. 

A question: What version of Node are you using?

**Comment by fflaten** [comment]
v18.16.0

**Comment by Princesseuh** [comment]
Ah, then I know the issue, Sharp's binaries lowest supported version is ^18.17.0. Try upgrading to at least 18.17.0 (the full semver check is `^18.17.0 || ^20.3.0 || >=21.0.0`)

That's unfortunate, wonder why it's that version specifically, we might have to pin until the next major of Astro... so annoying.

**Comment by fflaten** [comment]
I'm new to JS/TS world, but shouldn't it complain during `pnpm install` about that?
I get the same error with v18.19.0. Trying to reproduce in a new repo with devcontainer/codespace now.

**Comment by Princesseuh** [comment]
> I'm new to JS/TS world, but shouldn't it complain during `pnpm install` about that while installing `sharp`? 😕 Either way I still get the same error with v18.19.0. Trying to reproduce in a new repo with devcontainer/codespace now.

No, because those are all optionalDeps, the JS world doesn't handle this super well, unfortunately...

**Comment by fflaten** [comment]
> Either way I still get the same error with v18.19.0. Trying to reproduce in a new repo with devcontainer/codespace now.

Scratch that. Forgot to wipe node_modules after changing versions. v18.17.0+ works 🥳
Sorry for not testing that.

**Comment by PaulMorris** [comment]
I was getting a build failure on Cloudflare pages: `[ERROR] [vite] [vite]: Rollup failed to resolve import "sharp"` while using nodejs `18.14.1`. I found this issue, upgraded to nodejs `20.10.0` and that resolved the issue.

**Comment by LouisJS** [comment]
Astro v4.0.8
Node v20.10.0
System macOS (x64)
Package Manager npm
Output server
Adapter @astrojs/vercel/serverless
Integrations @astrojs/react
@astrojs/tailwind

Using npm run dev in development works. But as soon as i host it on Vercel : https://louis-lecocq.vercel.app/

**Comment by Boston343** [comment]
@fflaten yeah I've deleted node_modules, along with dist, and .astro folders at various points and it doesn't seem to help. 

Currently running Node v20.10.0 and npm 10.2.3, otherwise same stuff as original post

I'm getting intermittent errors now with Sharp even while in dev mode with Astro 4.0.8 where it says "Could not find Sharp.", when it could just 2 minutes ago, without me even closing the dev environment (just updating a file so it reloads). I have not managed to find consistency with what causes it. Right now I'm working with 4.0.7 and it seems fine. Seems like Sharp ^0.33.0 is just causing problems.

**Comment by emrerdem1** [comment]
This issue occurs only after migrating to `v4.1.0`.

Node: v20.10.0
Astro: 4.1.0

I'm getting these warnings about `Sharp` after nuking and re-installing `node_modules`:

```
warning package.json: No license field
warning landing-page@0.0.1: No license field
[1/4] Resolving packages...
[2/4] Fetching packages...
warning sharp@0.33.1: The engine "libvips" appears to
warning @img/sharp-darwin-arm64@0.33.1: The engine "p
warning @img/sharp-darwin-arm64@0.33.1: The engine "g
warning @img/sharp-darwin-x64@0.33.1: The engine "pnp
warning @img/sharp-darwin-x64@0.33.1: The engine "gli
warning @img/sharp-libvips-darwin-arm64@1.0.0: The en
warning @img/sharp-libvips-darwin-arm64@1.0.0: The en
warning @img/sharp-libvips-darwin-x64@1.0.0: The engi
warning @img/sharp-libvips-darwin-x64@1.0.0: The engi
warning @img/sharp-libvips-linux-arm@1.0.0: The engin
warning @img/sharp-libvips-linux-arm@1.0.0: The engin
warning @img/sharp-libvips-linux-arm64@1.0.0: The eng
warning @img/sharp-libvips-linux-arm64@1.0.0: The eng
warning @img/sharp-libvips-linux-s390x@1.0.0: The eng
warning @img/sharp-libvips-linux-s390x@1.0.0: The eng
warning @img/sharp-libvips-linux-x64@1.0.0: The engin
warning @img/sharp-libvips-linux-x64@1.0.0: The engin
warning @img/sharp-libvips-linuxmusl-arm64@1.0.0: The
warning @img/sharp-libvips-linuxmusl-arm64@1.0.0: The
warning @img/sharp-libvips-linuxmusl-x64@1.0.0: The e
warning @img/sharp-libvips-linuxmusl-x64@1.0.0: The e
warning @img/sharp-linux-arm@0.33.1: The engine "pnpm
warning @img/sharp-linux-arm@0.33.1: The engine "glib
warning @img/sharp-linux-arm64@0.33.1: The engine "pn
warning @img/sharp-linux-arm64@0.33.1: The engine "gl
ine "musl" appears to be invalid.
warning @img/sharp-wasm32@0.33.1: The engine "pnpm" appears to be invalid.
warning @img/sharp-win32-ia32@0.33.1: The engine "pnpm" appears to be invalid.
warning @img/sharp-win32-x64@0.33.1: The engine "pnpm" appears to be invalid.
[3/4] Linking dependencies...
warning "@astrojs/react > @vitejs/plugin-react@4.2.1" has unmet peer dependency "vite@^4.2.0 || ^5.0.0".
warning Workspaces can only be enabled in private projects.
[4/4] Building fresh packages...
```

Error:

```
Could not process image request: MissingSharp: Could not find Sharp. Please install Sharp (`sharp`) manually into your project or migrate to another image service.
    at loadSharp (C:/Users/A/project/node_modules/astro/dist/assets/services/sharp.js:19:11)
    at async Object.transform (C:/Users/A/project/node_modules/astro/dist/assets/services/sharp.js:31:15) 
    at async Module.GET (C:/Users/A/project/node_modules/astro/dist/assets/endpoint/node.js:67:30)        
    at async callEndpoint (file:///C:/A/A/project/node_modules/astro/dist/core/endpoint/index.js:124:16)
    at async #tryRenderRoute (file:///C:/Users/A/project/node_modules/astro/dist/core/pipeline.js:119:16) 
    at async DevPipeline.renderRoute (file:///C:/Users/A/project/node_modules/astro/dist/core/pipeline.js:58:20)
    at async handleRoute (file:///C:/Users/A/project/node_modules/astro/dist/vite-plugin-astro-server/route.js:248:18)
    at async run (file:///C:/Users/A/project/node_modules/astro/dist/vite-plugin-astro-server/request.js:50:14)
    at async runWithErrorHandling (file:///C:/Users/A/project/node_modules/astro/dist/vite-plugin-astro-server/controller.js:64:5)
    at async handleRequest (file:///C:/Users/A/project/node_modules/astro/dist/vite-plugin-astro-server/request.js:44:3) {
  loc: undefined,
  title: 'Could not find Sharp.',
  hint: "See Sharp's installation instructions for more information: https://sharp.pixelplumbing.com/install. If you are not relying on `astro:assets` to optimize, transform, or process any images, you can configure a passthrough image service instead 
of installing Sharp. See https://docs.astro.build/en/reference/errors/missing-sharp for more information.\n" +
    '\n' +
    'See https://docs.astro.build/en/guides/images/#default-image-service for more information on how to migrate to another image service.',
  frame: undefined,
  type: 'AstroError'
}
22:49:22 [500] /_image 130ms
```

`Sharp` dep is getting installed just fine in `v4.0.6`.

**Comment by Princesseuh** [comment]
4.1.0 includes no changes to this, so it's most likely a coincidence!

**Comment by emrerdem1** [comment]
> 4.1.0 includes no changes to this, so it's most likely a coincidence!

It could be. I tried upgrading to `4.1.0` twice, and had to downgrade back to `4.0.6`.  I don't know what's the cause, unfortunately.

**Comment by vrabe** [comment]
> It could be. I tried upgrading to `4.1.0` twice, and had to downgrade back to `4.0.6`. I don't know what's the cause, unfortunately.

Downgrading to `4.0.6` works for me though warnings still appear. I don't know why either.

**Comment by silveltman** [comment]
> The fix for me was to run yarn add sharp --ignore-engines, which seems to persist even when node_modules is removed. It's weird though, because I already had sharp in package.json, so I'd expect it to have worked. Computers 🤷 (edit: maybe I spoke too soon on that last point!)

`yarn add sharp --ignore-engines` also fixed it for me

**Comment by AdamKniec** [comment]
I had a problem to render any image on Astro v4.0.8 using the built in <Image> component...  I was about to give up but luckily I found this thread. 

If you are getting errors / warnings about missing Sharp or images beeing undefined even though the path you give to src is correct then try 

`yarn add sharp --ignore-engines`

I don't know if it's the proper way to fix that but I'm finally able to render the image... 

Thanks you good people on this thread :D

**Comment by joao-pedrozo** [comment]
> For me all tips here weren't enough, I had to update pnpm itself with `pnpm i --global pnpm` (as well as setting a recent node version)

Thanks

**Comment by jacklenox** [comment]
Just to confirm that for me as well, I just needed to update my versions of node and npm. I use nvm, so it was a matter of: `nvm install --lts`

**Comment by nemanjam** [comment]
> I had the same problem installing my site on my server using `@astrojs/node`, using Astro 4.0.3.
> 
> Installing `"sharp": "^0.32"` on the server (with node v20.10.0) solved the problem. On my computer (with node v21.2.0), `"sharp": "^0.33.1"` works fine.

`0.32.6` works, while `0.33.3` fails.

**Comment by EKliptus** [comment]
In my case, I solved this issue for build but not for astro dev ???

**TL;DR**
With astro, pnpm, node up-to date and sharp manual install v0.32.6 :
-> build on linux prod server : OK
-> build on windows dev laptop : OK
-> astro dev on the same windows dev laptop : KO

```
Astro               v4.8.6
Node                v20.13.1
System              Windows (x64)
Package Manager     pnpm (v9.1.1)
Output              static
Adapter             none
Integrations        @astrojs/sitemap
                    @astrojs/tailwind
                    @astrojs/alpinejs
                    @vite-pwa/astro-integration
```
**_Before this issue appear to me, I am using only `@vite-pwa/assets-generator` with `pwaAssets` integration that automatically generate pwa assets in dev and build mode._**

Then I wanted to use astro `assets:image` and I must manually install sharp :
- with sharp 0.33.x -> no build and no astro dev (error trigger by `@vite-pwa/assets-generator`).
- with sharp 0.32.6 -> builds are ok but not astro dev (error trigger by Astro)

With sharp 0.32.6 manual install :
- ✔when build no error appear (linux or windows) : astro assets:image and pwaAssets runs like a charm.
BUT
- ⁉ I have errors on pages with `assets:image Image` tag : **MissingSharp Astro Error** 
- `@vite-pwa/assets-generator` use sharp without error with astro dev in my case because I used `pwaAssets` integration

Finally when I deactivate pwaAssets integration from `@vite-pwa/assets-generator` all runs without error but I loose my pwa assets (404) in astro dev.

I see :
- astro `assets:image` must work with sharp 0.32.6
- when `@vite-pwa/assets-generator` works with up-to-date sharp 0.33.4

```
@astrojs/tailwind 5.1.0
└─┬ astro 4.8.6 peer
  └── sharp 0.33.4
@vite-pwa/assets-generator 0.2.4
├── sharp 0.32.6
└─┬ sharp-ico 0.1.5
  └── sharp 0.32.6
@vite-pwa/astro 0.3.1
├─┬ @vite-pwa/assets-generator 0.2.4 peer
│ ├── sharp 0.32.6
│ └─┬ sharp-ico 0.1.5
│   └── sharp 0.32.6
├─┬ astro 4.8.6 peer
│ └── sharp 0.33.4
└─┬ vite-plugin-pwa 0.19.8 peer
  └─┬ @vite-pwa/assets-generator 0.2.4 peer
    ├── sharp 0.32.6
    └─┬ sharp-ico 0.1.5
      └── sharp 0.32.6
astro 4.8.6
└── sharp 0.33.4
sharp 0.32.6
sharp-ico 0.1.5
└── sharp 0.32.6
vite-plugin-pwa 0.19.8
└─┬ @vite-pwa/assets-generator 0.2.4 peer
  ├── sharp 0.32.6
  └─┬ sharp-ico 0.1.5
    └── sharp 0.32.6
```

Thank you for your clarifications.

**Comment by EKliptus** [comment]
For the moment, my workaround are :
- either I bypass Astro `assets:image` Image` tag processing with `passthroughImageService()` in dev mode
- either I disable `pwaAssets` integration in AstroPWA integration in dev mode

**Comment by Princesseuh** [comment]
It seems like the PWA plugin you're using is using an outdated version of Sharp perhaps? Recent versions of Astro assume Sharp 0.33.x

**Comment by EKliptus** [comment]
> It seems like the PWA plugin you're using is using an outdated version of Sharp perhaps? Recent versions of Astro assume Sharp 0.33.x
Ok yes it is ! brain fog... It's more clear for me now...
Thank you @Princesseuh

**Comment by dfrios** [comment]
> I had a problem to render any image on Astro v4.0.8 using the built in `<Image>` component... I was about to give up but luckily I found this thread.
> 
> If you are getting errors / warnings about missing Sharp or images beeing undefined even though the path you give to src is correct then try
> 
> `yarn add sharp --ignore-engines` just like guys above suggested
> 
> I don't know if it's the proper way to fix that but I'm finally able to render the image...
> 
> Thanks you good people on this thread :D

This worked for me.

**Comment by bskimball** [comment]
For me to fix:

- upgrade Node to 20.16
- remove node_modules folder
- npm install

**Comment by Giovannisabces** [comment]
For me to fix:

- update Node to v20.17.0.
- install sharp becouse i use pnpm.
- remove node_modules folder.
- pnpm install.

**Comment by JordanMcn** [comment]
> For me to fix:
> 
> * update Node to v20.17.0.
> * install sharp becouse i use pnpm.
> * remove node_modules folder.
> * pnpm install.

Thanks @Giovannisabces ! This worked for me too

**Comment by moojing** [comment]
I still encountered this `CouldNotTransformImage` and sharp not found error at `astro@4.8.6`, until I add sharp into my dependencies by using `pnpm add shart` and it's solved now. It's weird but thanks to the guys here for mentioning the solution!

---------------
Astro                    v4.8.6
Node                     v20.10.0
System                   MacOS
Package Manager          pnpm
Output                   static
---------------

**Comment by vladinator1000** [comment]
For me Sharp works locally but fails when building in github actions using the standard starlight template.

**Comment by wiki-Bird** [comment]
I'm getting the same as @vladinator1000, except on cloudflare pages rather than gh actions

**Comment by xamgore** [comment]
> by using `pnpm add shart`

not t but p `pnpm add sharp`
